homeassistant/components/onvif/__init__.py

```diff
@@ -155,3 +158,49 @@ async def async_populate_options(hass: HomeAssistant, entry: ConfigEntry) -> Non
     }
 
     hass.config_entries.async_update_entry(entry, options=options)
+
+
+@callback
+def _migrate_camera_entities_unique_ids(
+    hass: HomeAssistant, config_entry: ConfigEntry, device: ONVIFDevice
+) -> None:
+    """Migrate unique ids of camera entities from profile index to profile token."""
+    entity_reg = er.async_get(hass)
+    entities: list[er.RegistryEntry] = er.async_entries_for_config_entry(
+        entity_reg, config_entry.entry_id
+    )
+
+    mac_or_serial = device.info.mac or device.info.serial_number
+    old_uid_start = f"{mac_or_serial}_"
+    new_uid_start = f"{mac_or_serial}#"
+
+    for entity in entities:
+        if entity.domain != Platform.CAMERA:
+            continue
+
+        if (
+            not entity.unique_id.startswith(old_uid_start)
+            and entity.unique_id != mac_or_serial
+        ):
+            continue
+
+        index = 0
+        if entity.unique_id.startswith(old_uid_start):
+            try:
+                index = int(entity.unique_id[len(old_uid_start) :])
+            except ValueError:
+                LOGGER.error(
+                    "Failed to migrate unique id for '%s' as the ONVIF profile index could not be parsed from '%s'",
+                    entity.entity_id,
+                    entity.unique_id,
+                )
+                continue
+        token = device.profiles[index].token
```

Comment:
I think its possible that the profile is no longer available or has been disabled on the camera since the entities were created which could cause this to raise `IndexError`

Reply:
I'll handle it.

Reply:
Done.